### PR TITLE
avm1: Fix coercion order in SWFv4 actions

### PR DESCRIPTION
**Pull request by Toad06**
I also added a SWF that tests the behavior for almost all of the SWFv4 actions when dealing with coercion.
This test would fail if run in the current builds:

test

### CODE REVIEW

File: core/src/avm1/activation.rs [hunk 1]
```diff
@@ -650,8 +650,8 @@ impl<'a, 'gc> Activation<'a, 'gc> {
     }
 
     fn action_clone_sprite(&mut self) -> Result<FrameControl<'gc>, Error<'gc>> {
-        let depth = self.context.avm1.pop();
-        let target = self.context.avm1.pop();
+        let depth = self.context.avm1.pop().coerce_to_f64(self)?;
```
Comment (relrelb) [inline core/src/avm1/activation.rs:653]: While reading related code, I noticed some more inaccuracies around `ActionCloneSprite` and `duplicateMovieClip()`. Opened #11228 for that.

File: core/src/avm1/activation.rs [hunk 2]
```diff
@@ -1009,9 +1009,9 @@ impl<'a, 'gc> Activation<'a, 'gc> {
     }
 
     fn action_end_drag(&mut self) -> Result<FrameControl<'gc>, Error<'gc>> {
-        // we might not have had an opportunity to call `update_drag`
-        // if AS did `startDrag(mc);stopDrag();` in one go
-        // so let's do it here
+        // We might not have had an opportunity to call `update_drag`
```
Comment (relrelb) [inline core/src/avm1/activation.rs:1012]: nit: This comment appears in two more places - `core/src/avm1/globals/movie_clip.rs` and `core/src/avm2/globals/flash/display/sprite.rs`. Might be worth to update them as well.

File: core/src/avm1/activation.rs [hunk 3]
```diff
@@ -1883,22 +1890,43 @@ impl<'a, 'gc> Activation<'a, 'gc> {
...
+
+        let (property, prop_value) = if !prop_index.is_finite() || prop_index <= -1.0 {
+            (None, prop_value)
+        } else {
+            let display_properties = self.context.avm1.display_properties();
+            let properties = display_properties.read();
+            if let Some(property) = properties.get_by_index(prop_index as usize) {
+                // `prop_value` is coerced here, as the coercion must happen even if the target
+                // is invalid or the property is read-only.
+                let coerced_value = crate::avm1::object::stage_object::action_property_coerce(
```
[the remark this answers is not in the document]
Comment (Toad06) [reply core/src/avm1/activation.rs:1905]: Your sample traces "!!!" on my end (made with JPEXS and output checked from the debug projector):
[sample.zip](https://github.com/ruffle-rs/ruffle/files/11582481/sample.zip)

Maybe the Flash editor was able to detect that the target was invalid when compiling, and so completely removed the call to setProperty from the final SWF? (this is pure speculation 😄)

[the remark this answers is not in the document]
Comment (Dinnerbone) [reply core/src/avm1/activation.rs:1899]: It looks like this is player-version specific, not swf-version. In 32 I get !!!, in 11 I get nothing. I don't know when it changed, but Ruffle makes it always print right now despite the player version.

[the remark this answers is not in the document]
Comment (Lord-McSweeney) [reply core/src/avm1/activation.rs:1899]: The behavior should probably be changed based on `avm1.player_version`, then.

[the remark this answers is not in the document]
Comment (Toad06) [reply core/src/avm1/activation.rs:1899]: Good finding. Here's the different behaviors I observed in the Flash Player versions I tested:
- Flash Player 11 and below + FP 15: Only coerces if target is valid. 
- Flash Player 18: Coerces even if target is invalid.
- Flash Player 19-20: No coercion at all (even if target is valid).
- Flash Player 21 to 32: Coerces even if target is invalid.

Unless for significant breaking changes (like the NaN equality), we should only focus on reproducing the behavior from FP 32 imho.

**Review by Dinnerbone — APPROVED** [review]

**Pull request merged**